pallets/pips/src/lib.rs

```diff
@@ -742,126 +734,79 @@ decl_module! {
+            // 1. Proposal must be from the community.
+            let proposer = match meta.proposer {
+                Proposer::Committee(_) => return Err(Error::<T>::NotFromCommunity.into()),
+                Proposer::Community(p) => p,
+            };
 
-            // Valid PipId
-            ensure!(<ProposalResult<T>>::contains_key(id), Error::<T>::NoSuchProposal);
+            if proposer == voter {
+                // 2a. Deposit must be above minimum.
+                // Note that proposer can still vote against their own PIP.
+                ensure!(deposit >= Self::min_proposal_deposit(), Error::<T>::IncorrectDeposit);
+            } else {
+                // 2b. Only proposer can vote during PIP's cool-off period.
+                let curr_block_number = <system::Module<T>>::block_number();
+                ensure!(meta.cool_off_until <= curr_block_number, Error::<T>::ProposalOnCoolOffPeriod);
+            }
 
-            // Double-check vote duplication.
-            // TODO(centril): Do we really want to prevent duplicate votes -- why not aggregate instead?
-            ensure!(matches!(Self::proposal_vote(id, &proposer), None), Error::<T>::DuplicateVote);
+            // 3. Proposal must be pending.
+            Self::is_proposal_state(id, ProposalState::Pending)?;
 
-            // Reserve the deposit
-            <T as Trait>::Currency::reserve(&proposer, deposit).map_err(|_| Error::<T>::InsufficientDeposit)?;
+            // TODO(centril): move this to a suitable utils crate.
+            fn with_transaction<T, E>(tx: impl FnOnce() -> Result<T, E>) -> Result<T, E> {
+                use frame_support::storage::{with_transaction, TransactionOutcome};
+                with_transaction(|| match tx() {
+                    r @ Ok(_) => TransactionOutcome::Commit(r),
+                    r @ Err(_) => TransactionOutcome::Rollback(r),
+                })
+            }
 
-            // Save your vote.
-            Self::unsafe_vote(id, proposer.clone(), Vote(aye_or_nay, deposit))
-                .map_err( |vote_error| {
-                    debug::warn!("The counters of voting (id={}) have an overflow, transaction is roll-back", id);
-                    let _ = <T as Trait>::Currency::unreserve(&proposer, deposit);
-                    vote_error
-                })?;
+            with_transaction(|| {
+                // 4. Reserve the deposit, or refund if needed.
+                let curr_deposit = Self::deposits(id, &voter).amount;
+                if deposit < curr_deposit {
+                    <T as Trait>::Currency::unreserve(&voter, curr_deposit - deposit);
+                } else {
+                    <T as Trait>::Currency::reserve(&voter, deposit - curr_deposit).map_err(|_| Error::<T>::InsufficientDeposit)?;
+                }
+                // 5. Save the vote.
+                Self::unsafe_vote(id, voter.clone(), Vote(aye_or_nay, deposit))
+            })?;
```

Comment:
Why is rollback required here? I don't see a scenario where these functions modify storage and then throw an error.

Reply:
`Currency::reserve` can modify storage and then `Self::unsafe_vote` can throw `Error::StakeAmountOfVotesExceeded`.
This transaction was introduced to handle a test failure.